[Merged by Bors] - Drop p2p connections on ValidationReject

piersy opened the pull request:
## Motivation
We don't want to stay connected to peers that we know are exhibiting invalid behaviour.
This is a step towards completing #4088

## Changes
Change the signature of `pubsub.GossipHandler` to return an error instead of a `pubsub.ValidationResult` this makes the interface used by both sync handlers and pubsub handlers consistent, which allows for the use of a single decorator that will drop a peer if the handler returns a specific error. The decorator (`pubsub.DropPeerOnError`) is applied to all pubsub handlers and sync handlers.

Added a test to verify disconnection on pubsub handlers, to support this I exposed the `p2p.Host` from `node.App`

## Test Plan
The test shows that this is working. A test or tests to verify this behaviour for sync handlers has yet to be added

## DevOps Notes
<!-- Please uncheck these items as applicable to make DevOps aware of changes that may affect releases -->
- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

piersy committed:
Drop p2p connections on ValidationReject

piersy committed:
Extend test of peer disconnection

Try to verify more precisely that only messages the result in a
ValidationReject outcome result in disconnection.

dshulyak commented on p2p/pubsub/wrapper.go:
```diff
@@ -34,6 +36,19 @@ func (ps *PubSub) Register(topic string, handler GossipHandler) {
+			// Ok we close the peer here but the deadpeerhandler may try a redial if the state of the conn is connected
+			err := ps.host.Network().ClosePeer(pid)
+			if err != nil {
+				ps.logger.With().Error("failed to close peer",
```
this should be Debug. for information maybe it makes sense to add a counter how many peers are disconnected

but also this approach is not sufficient, as mentioned we need to drop peer also when it gives bad message over sync. thats why i suggested to decorate handlers

piersy replied:
After discussion regarding how to handle bad sync messages, it was decided that handlers should return errors as opposed to validation results and then both sync and pubsusb handlers can share a single wrapper that handles peer dropping.

piersy replied:
@dshulyak I've made the changes we discussed, please take a look.

piersy committed:
Merge branch 'develop' into piersy/drop-p2p-connections-on-validation-reject

piersy committed:
Update test to set individual file locks

piersy committed:
Handlers return error instead of validation result

This is a step towards unifying the function signatures for sync
handlers and pubsub hanlders.

This compiles but some tests are probably broken.

piersy committed:
Fix tests broken by handler change

piersy committed:
Reinstate logging of specific handler errors

piersy committed:
Add DropPeerOnError GossipHandler decorator

The decorator is applied to both the pubsub handlers and the sync
handlers.

piersy committed:
Reinstate comment for GossipHandler

piersy committed:
Merge remote-tracking branch 'origin/develop' into piersy/drop-p2p-connections-on-validation-reject

dshulyak commented on p2p/pubsub/pubsub.go:
```diff
@@ -139,15 +141,40 @@ const (
 	ValidationReject = pubsub.ValidationReject
 )
```
i think we should remove all aliases above

piersy replied:
Do you know the reason for adding them in the first place, they seemed kind of unnecessary to me?

dshulyak replied:
i defined aliases so that libp2p is not imported directly across project codebase, it is easier to prevent misuse this way

dshulyak commented on p2p/pubsub/pubsub.go:
```diff
@@ -139,15 +141,40 @@ const (
+	return func(ctx context.Context, peer peer.ID, data []byte) error {
+		err := handler(ctx, peer, data)
+		if errors.Is(err, targetErr) {
+			// We want to disconnect the peer and also penalize it which could result in it being blacklisted.
```
imo this comment is not that useful

dshulyak commented on p2p/pubsub/pubsub.go:
```diff
@@ -139,15 +141,40 @@ const (
+				logger.With().Debug("failed to close peer",
+					log.String("peer", peer.ShortString()),
+					log.Err(err),
+				)
```
we should also track how many peers we are dropping. maybe just a prometheus counter here

dshulyak commented on p2p/pubsub/pubsub.go:
```diff
@@ -139,15 +141,40 @@ const (
+// pubsub validation result is ValidationReject. ValidationAccept is indicated
+// by a nil error and ValidationIgnore is indicated by any error that is not a
+// ValidationRejectErr.
+var ValidationRejectErr = errors.New("validation reject")
```
as linter said this should be ErrValidationReject, maybe ErrReject

dshulyak commented on p2p/bad_peer_test.go:
```diff
@@ -0,0 +1,154 @@
+	g.Go(func() error {
+		return app1.Start(ctx)
+	})
+	<-app1.Started()
```
would be better if this couldn't get stuck for whole test timeout if app can't start

dshulyak replied:
it is also better to use mocknet from libp2p. tcp stack was very flaky on ci runners in the past

piersy replied:
Some of the configuration is done in node, so to test that we need to use node. I think it's nice also to know that the peer dropping is actually working in an approximately real system as opposed to a much more contained test that would involve the mocknet.

I'm interested to know more about the flaky tcp stack, I've never heard of a tcp stack being flaky?

dshulyak replied:
it is environment issue, i am pretty sure that you will see it again with this test

piersy replied:
>it is environment issue, i am pretty sure that you will see it again with this test

Maybe this has been resolved? I've managed to run this test 23000 times on CI without a single failure.

dshulyak commented on p2p/bad_peer_test.go:
```diff
@@ -0,0 +1,154 @@
+	"go.uber.org/zap/zapcore"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/spacemeshos/go-spacemesh/cmd/node"
```
i don't think that we should import node here. is it possible to write a test without a node, or otherwise add this test to a node module

piersy replied:
We can add it to the node module, but what is the problem you see with importing node here?

dshulyak replied:
~you will see a cycle if you want to import this module in the tests for node~ test in *_test.go, but i still prefer to avoid node imports here.

and even conceptually, there should be a hierarchy of imports in the project. higher level things (as node) should not be imported by lower level things (such as encoding, p2p, database)

i think it will also lead to failures in p2p module, because of some changes in node. in my opinion thats a mess

piersy replied:
With regards to the hierarchy totally agree with respect to production code. But for tests I think this can be a good approach.

Consider a situation where you want to test an API and you wan't to test it with the full stack active, I.E not mocking out any stuff so having a full node running. You could put these tests in node, but does it really make sense to put the tests in node when they are testing an API that is defined in a different package? If you do that for all the APIs the node package will end up with a lot of tests that don't really have anything to do with node. Structurally there is no problem with importing node in test packages because you can't depend on a test package.

> i think it will also lead to failures in p2p module, because of some changes in node. in my opinion thats a mess

I don't see this as a problem, if node is not supporting correct functioning of the API then that's a real problem. For example if the API is not registered in node then it's not going to work, that is something we want to know about.

piersy replied:
I've moved the test to the node package.

dshulyak commented on cmd/node/node.go:
```diff
@@ -749,19 +749,28 @@ func (app *App) initServices(
 	)
-	fetcher.SetValidators(atxHandler, poetDb, proposalListener, blockHandler, proposalListener, app.txHandler, malfeasanceHandler)
+	fetcher.SetValidators(
+		fetch.ValidatorFunc(pubsub.DropPeerOnError(atxHandler.HandleAtxData, app.host, lg, pubsub.ValidationRejectErr)),
```
can we maybe do DropPeerOnReject (as a wrapper around DropPeerOnError, if you want to keep that)?

piersy replied:
I thought it could be useful to make it more general, but If you don't think so I'll just make it `DropPeerOnValidationReject` and hardcode the error.

dshulyak replied:
i meant that you can keep both of them, but define "shortcut" for Reject, because it is used everywhere

piersy replied:
I din't feel like keeping the original implementation and the shortcut, it's easy enough to split out if we need to.  Having both added an extra layer of indirection that I felt was not necessary.

I would have been happy just with the more general version though, IMO the shortcut doesn't provide enough value to justify it's use over the more general version.

EG.

```
pubsub.DropPeerOnError(atxHandler.HandleAtxData, app.host, lg, pubsub.ValidationRejectErr)
```
VS
```
pubsub.DropPeerOnValidationRejectError(atxHandler.HandleAtxData, app.host, lg)
```

But I don't feel that strongly about this.

dshulyak commented on blocks/handler.go:
```diff
@@ -9,13 +9,14 @@ import (
 
 var (
-	errMalformedData  = errors.New("malformed data")
+	errMalformedData  = fmt.Errorf("malformed data: %w", pubsub.ValidationRejectErr)
```
does it make more sense to have message like: "reject: some reason", so the template in this case "%w: malformed data"?

piersy replied:
Yes it does, thanks!

piersy committed:
Reword rejection errors

"malformed data: reject" -> "reject: malformed data"

piersy committed:
Remove p2p rejection result aliases

piersy committed:
Remove unuseful comment

piersy committed:
Rename ValidationRejectErr -> ErrValidationReject

piersy committed:
Hardcode rejection error into drop peer decorator

piersy committed:
Added test of connection dropping from the fetcher

It is failing randomly though, it seems that sometimes even though
the connection is being closed it is not actually being disconnected.

piersy committed:
Switched to using real hosts for fetcher test

The real hosts seem to be able to have connections closed reliably.

piersy committed:
Clean up  fetcher test

piersy committed:
Extend fetcher peer dropping test

It now validates that a without a peer dropping validator peers are not
dropped.

piersy committed:
Move bad_peer_test.go to node package

piersy committed:
Remove debug print statement

piersy committed:
Add counter to track dropped connections

piersy committed:
Merge branch 'develop' into piersy/drop-p2p-connections-on-validation-reject

piersy committed:
Fix compile error due to params change in merge

piersy committed:
Copy genesis config across

dshulyak commented:
there is this failure

![Uploading image.png…]()

piersy committed:
Ensure protocol prefix is applied

piersy committed:
Revert "Ensure protocol prefix is applied"

This reverts commit 8504fae4c8ebc6b0c1705d2599496fd1a9c438df.

piersy committed:
Merge branch 'develop' into piersy/drop-p2p-connections-on-validation-reject

piersy committed:
Revert "Fix compile error due to params change in merge"

This reverts commit f7f2bbde28ca89cf0ee770d5d7bb08f2824ecf26.

piersy commented:
bors merge

dshulyak reviewed (commented):
```
diff --git a/cmd/node/node.go b/cmd/node/node.go
index 318d1542d..733b5ef58 100644
--- a/cmd/node/node.go
+++ b/cmd/node/node.go
@@ -763,7 +763,7 @@ func (app *App) initServices(
        fetcher.SetValidators(
                fetch.ValidatorFunc(pubsub.DropPeerOnValidationReject(atxHandler.HandleAtxData, app.host, lg)),
                fetch.ValidatorFunc(pubsub.DropPeerOnValidationReject(poetDb.ValidateAndStoreMsg, app.host, lg)),
-               fetch.ValidatorFunc(pubsub.DropPeerOnValidationReject(proposalListener.HandleSyncedProposal, app.host, lg)),
+               fetch.ValidatorFunc(pubsub.DropPeerOnValidationReject(proposalListener.HandleSyncedBallot, app.host, lg)),
```

there was a typo

piersy committed:
Fix typo, ensure fetcher validates types correctly

piersy commented:
bors merge
